src/Disks/IO/CachedReadBufferFromRemoteFS.cpp

```diff
@@ -282,29 +282,28 @@ SeekableReadBufferPtr CachedReadBufferFromRemoteFS::getReadBufferForFileSegment(
             }
             case FileSegment::State::PARTIALLY_DOWNLOADED:
             {
-                auto downloader_id = file_segment->getOrSetDownloader();
-                if (downloader_id == file_segment->getCallerId())
-                {
-                    size_t download_offset = file_segment->getDownloadOffset();
-                    bool can_start_from_cache = download_offset > file_offset_of_buffer_end;
+                size_t download_offset = file_segment->getDownloadOffset();
+                bool can_start_from_cache = download_offset > file_offset_of_buffer_end;
 
-                    LOG_TEST(log, "Current download offset: {}, file offset of buffer end: {}", download_offset, file_offset_of_buffer_end);
+                LOG_TEST(log, "Current download offset: {}, file offset of buffer end: {}", download_offset, file_offset_of_buffer_end);
 
-                    if (can_start_from_cache)
-                    {
-                        ///                      segment{k}
-                        /// cache:           [______|___________
-                        ///                         ^
-                        ///                         download_offset
-                        /// requested_range:    [__________]
-                        ///                     ^
-                        ///                     file_offset_of_buffer_end
+                if (can_start_from_cache)
+                {
+                    ///                      segment{k}
+                    /// cache:           [______|___________
+                    ///                         ^
+                    ///                         download_offset
+                    /// requested_range:    [__________]
+                    ///                     ^
+                    ///                     file_offset_of_buffer_end
 
-                        read_type = ReadType::CACHED;
-                        file_segment->resetDownloader();
-                        return getCacheReadBuffer(range.left);
-                    }
+                    read_type = ReadType::CACHED;
+                    return getCacheReadBuffer(range.left);
+                }
 
+                auto downloader_id = file_segment->getOrSetDownloader();
```

Comment:
This would not be fully correct.
It is possible that `can_start_from_cache` is `false`, then it can become `true` right before calling `file_segment->getOrSetDownlaoder()` which can also turn out as `true`. As a result we will use a buffer which reads from remote fs, but in fact should start with cached buffer. So we anyway need this `can_start_from_cache` check after `if (downloader_id == file_segment->getCallerId())`. 

Let's leave this `can_start_from_cache` check in both places.

Reply:
I got it, I will change my code. please review later.